### PR TITLE
gdal: Split tests into separate attributes

### PR DESCRIPTION
**Pull request by l0b0**
## Description of changes

Makes it easier to run single tests for debugging purposes.

Preparation for adding more tests in #290292.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [x] Tested, as applicable: (`nix-build -A gdal.passthru.tests`)
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by dotlambda — COMMENTED** [review]
Why not put the multiple attributes in `tests.nix`?

File: pkgs/development/libraries/gdal/default.nix [hunk 1]
```diff
@@ -270,8 +271,60 @@ stdenv.mkDerivation (finalAttrs: {
     popd # autotest
   '';
 
-  passthru.tests = {
-    gdal = callPackage ./tests.nix { gdal = finalAttrs.finalPackage; };
+  passthru.tests = let
+    gdal = finalAttrs.finalPackage;
+  in {
+    ogrinfo-gdal-version = runCommand "${gdal.pname}-ogrinfo-gdal-version" { } ''
+      ${gdal}/bin/ogrinfo --version | grep 'GDAL ${gdal.version}'
```
Comment (dotlambda) [inline pkgs/development/libraries/gdal/default.nix:278]: I like
```suggestion
      ${lib.getExe' gdal "ogrinfo"} --version | grep 'GDAL ${gdal.version}'
```
but it's up to you.

File: pkgs/development/libraries/gdal/default.nix [hunk 2]
```diff
@@ -270,8 +271,60 @@ stdenv.mkDerivation (finalAttrs: {
...
+      touch $out
+    '';
+    gdalinfo-gdal-version = runCommand "${gdal.pname}-gdalinfo-gdal-version" { } ''
```
Comment (dotlambda) [inline pkgs/development/libraries/gdal/default.nix:281]: This should use `testers.testVersion`.

File: pkgs/development/libraries/gdal/default.nix [hunk 3]
```diff
@@ -270,8 +271,60 @@ stdenv.mkDerivation (finalAttrs: {
...
+    java-bindings = runCommand "${gdal.pname}-java-bindings" { } ''
+      cat <<EOF > main.java
+      import org.gdal.gdal.gdal;
+      class Main {
+        public static void main(String[] args) {
+        gdal.AllRegister();
+        }
+      }
+      EOF
+      ${jdk}/bin/java -Djava.library.path=${gdal}/lib/ -cp ${gdal}/share/java/gdal-${gdal.version}.jar main.java
```
Comment (dotlambda) [inline pkgs/development/libraries/gdal/default.nix:325]:
```suggestion
      ${lib.getExe jdk} -Djava.library.path=${gdal}/lib/ -cp ${gdal}/share/java/gdal-${gdal.version}.jar main.java
```

File: pkgs/development/libraries/gdal/default.nix (hunk 1 above)
Comment (l0b0) [reply pkgs/development/libraries/gdal/default.nix:278]: Isn't `lib.getExe'` (with the apostrophe) an internal function? Is it actually meant to be used like that? If so, I'm all for it.

Comment (dotlambda) [reply pkgs/development/libraries/gdal/default.nix]: No it's already used quite a bit in nixpkgs.

**Review by imincik — APPROVED** [review]
I am happy with this change, even though I haven't seen this pattern anywhere else (do you know about any other package splitting tests in similar way ?) Also, more use of testers and lib functions is good thing as well.

**Comment by l0b0** [comment]
> I am happy with this change, even though I haven't seen this pattern anywhere else (do you know about any other package splitting tests in similar way ?)

I don't know where I picked it up (The documentation? Some other random package?), but looking for packages with more than one test in `nix repl`:

```
pkgs = import <nixpkgs> {}
packageTests = pkg: if builtins.isList (pkg.passthru.tests or {}) then {} else pkg.passthru.tests or {}
testNames = pkg: builtins.attrNames (packageTests pkg)
testCount = pkg: builtins.length (testNames pkg)
multiTest = pkg: let result = builtins.tryEval (testCount pkg); in if result.success then result.value > 1 else false
multiTests = pkgs.lib.filterAttrs (_name: pkg: multiTest pkg) pkgs
builtins.attrNames multiTests
```

This finds a bunch:

```
[ "agda" "akkoma" "alice-lg" "amazon-ssm-agent" "angie" "angieQuic" "apacheHttpd" "apacheHttpd_2_4" "avahi" "avahi-compat" "aws-sdk-cpp" "ayatana-indicator-messages" "azure-static-sites-client" "bats" "bazel" "bazel_4" "bazel_5" "bazel_6" "bazel_7" "bcachefs-tools" "bind" "bird" "bird2" "bitwarden_rs" "bitwarden_rs-mysql" "bitwarden_rs-postgresql" "bitwarden_rs-sqlite" "bitwarden_rs-vault" "boehmgc" "btrbk" "buck2" "c-ares" "c-aresMinimal" "cacert" "caddy" "cargo" "cargo-auditable-cargo-wrapper" "centrifugo" "ceph" "ceph-dev" "cf-terraforming" "checksec" "chrony" "cimg" "clevis" "cloud-init" "cntr" "conceal" "conmon" "containerd" "convos" "corretto11" "corretto17" "corretto19" "cpio" "craftos-pc" "ctranslate2" "cuneiform" "cups" "curl" "curlFull" "curlHTTP3" "curlMinimal" "curlWithGnuTls" "cyrus_sasl" "dart" "dart-sass" "dart_stable" "dav1d" "dconf2nix" "deno" "dex-oidc" "dhcpcd" "dig" "discourse" "discourseAllPlugins" "dnsmasq" "dnsutils" "dotnet-aspnetcore" "dotnet-aspnetcore_6" "dotnet-aspnetcore_7" "dotnet-aspnetcore_8" "dotnet-netcore" "dotnet-runtime" "dotnet-runtime_6" "dotnet-runtime_7" "dotnet-runtime_8" "dotnet-sdk" "dotnet-sdk_6" "dotnet-sdk_7" "dotnet-sdk_8" "dovecot" "drawterm" "drawterm-wayland" "dump_syms" "endlessh" "endlessh-go" "engelsystem" "envoy" "etcd" "etcd_3_5" "expat" "faad2" "faircamp" "faiss" "faissWithCuda" "fakeroot" "fcft" "fend" "ferretdb" "firejail" "fish" "flatpak" "fmt" "fmt_10" "fmt_8" "fmt_9" "foot" "forgejo" "fq" "frankenphp" "freetube" "freetype" "freshrss" "fwupd" "g4py" "garage" "garage_0_8" "garage_0_8_5" "garage_0_9" "garage_0_9_1" "gdk-pixbuf" "geant4" "geos" "geos_3_11" "ghostscript" "ghostscriptX" "ghostscript_headless" "ghostunnel" "ghz" "git" "git-doc" "gitFull" "gitMinimal" "gitSVN" "gitea" "glib" "gmic" "gmic-qt" "gns3-server" "gnupg" "gnupg22" "gnupg24" "gnutls" "go" "go_1_20" "go_1_21" "go_1_22" "goss" "gpgme" "grafana-agent" "graphene-hardened-malloc" "graphviz" "graphviz-nox" "groonga" "grpc" "grpc-gateway" "grub2" "grub2_efi" "grub2_full" "grub2_light" "grub2_xen" "hadoop" "hadoop2" "hadoop3" "hadoop_3_2" "hadoop_3_3" "handbrake" "harfbuzz" "harfbuzzFull" "hello" "hercules-ci-agent" "hockeypuck" "hockeypuck-web" "home-assistant" "host" "hotdoc" "htmldoc" "icloudpd" "idris2" "imagemagick" "imagemagick7" "imagemagick7Big" "imagemagick7_light" "imagemagickBig" "imagemagick_light" "imlib2" "imlib2-nox" "imlib2Full" "innernet" "input-remapper" "installShellFiles" "intel-gmmlib" "ipfs" "jenkins" "jj" "jupyter" "jupyter-all" "k3s" "k3s_1_24" "k3s_1_25" "k3s_1_26" "k3s_1_27" "k3s_1_28" "k3s_1_29" "kbd" "kcov" "kea" "keepassxc" "kerberos" "keycloak" "knot-dns" "krb5" "krb5Full" "kubectl" "kubectl-convert" "kubernetes" "kubo" "leetcode-cli" "libarchive" "libbpf" "libbpf_1" "libcap" "libcbor" "libceph" "libde265" "libdeltachat" "libgcrypt" "libgcrypt_1_8" "libgit2" "libgit2_1_3_0" "libgit2_1_5" "libgit2_1_6" "libheif" "libimagequant" "libjpeg" "libjpeg8" "libjpeg_turbo" "libkrb5" "libmad" "libnghttp2" "libpcap" "libraw" "librsvg" "libsass" "libsndfile" "libssh" "libtasn1" "libtiff" "libtransmission" "libtransmission_4" "libtree" "libuv" "libva" "libva-minimal" "libvmaf" "libwebp" "libxcrypt" "libxcrypt-legacy" "linux-pam" "lua" "lua5" "lua5_1" "lua5_2" "lua5_2_compat" "lua5_3" "lua5_3_compat" "lua5_4" "lua5_4_compat" "luajit" "luajit_2_0" "luajit_2_1" "luajit_openresty" "lvm2" "lvm2_dmeventd" "lvm2_vdo" "lxd" "lxd-unwrapped" "makeBinaryWrapper" "makeShellWrapper" "makeWrapper" "mariadb" "mariadb-client" "mariadb-embedded" "mariadb_1011" "mariadb_105" "mariadb_106" "mariadb_110" "matomo" "matomo-beta" "matrix-synapse-unwrapped" "mediawiki" "mercurial" "mercurialFull" "merecat" "miniflux" "miniupnpd" "miniupnpd-nftables" "mkl" "mold" "mupdf" "mupdf-headless" "musescore" "mysql" "mysql-client" "mysql80" "n8n" "nanopb" "nanopbMalloc" "neovim" "netbird" "netbird-ui" "netbox" "netbox_3_6" "netbox_3_7" "nextcloud-notify_push" "nextcloud26" "nextcloud27" "nextcloud28" "nfs-utils" "nftables" "nghttp2" "nginx" "nginxMainline" "nginxQuic" "nginxShibboleth" "nginxStable" "ngtcp2-gnutls" "nix" "nix-serve" "nixFlakes" "nixStable" "nixStatic" "nixfmt-rfc-style" "nixos-container" "nixos-install-tools" "nixos-rebuild" "nspr" "ocsinventory-agent" "octoprint" "omnisharp-roslyn" "open-watcom-bin" "open-watcom-v2" "open62541" "openblas" "openblasCompat" "opencl-headers" "opencv" "opencv4" "opencv4WithoutCuda" "openresty" "opensearch" "opensmtpd" "opensnitch" "openvpn" "orc" "pam" "parsedmarc" "pdal" "pdns-recursor" "percona-server_8_0" "pg_repack" "phoc" "php" "php81" "php82" "php83" "pinnwand" "pixman" "pjsip" "playwright" "please-cli" "plv8" "podman" "polaris" "pomerium" "poppler" "poppler_gi" "poppler_min" "poppler_utils" "postfix" "postgresql" "postgresql_12" "postgresql_12_jit" "postgresql_13" "postgresql_13_jit" "postgresql_14" "postgresql_14_jit" "postgresql_15" "postgresql_15_jit" "postgresql_16" "postgresql_16_jit" "postgresql_jit" "powerdns-admin" "prefetch-npm-deps" "prefetch-yarn-deps" "proj" "prometheus-php-fpm-exporter" "prosody" "protobuf" "protobuf3_21" "protobuf3_23" "protobuf3_24" "protobuf_21" "protobuf_23" "protobuf_24" "protobuf_25" "pufferpanel" "pypy" "pypy2" "pypy27" "pypy3" "pypy310" "pypy39" "pyrosimple" "python3" "python310" "python310Full" "python311" "python311Full" "python312" "python312Full" "python313" "python313Full" "python38" "python38Full" "python39" "python39Full" "python3Full" "python3Minimal" "qemu" "qemu_full" "qemu_kvm" "qemu_test" "qemu_xen" "qemu_xen-light" "qemu_xen_4_15" "qemu_xen_4_15-light" "qrencode" "quartus-prime-lite" "quickemu" "rasdaemon" "re2" "re2c" "repro-get" "resholve" "retroarchBare" "ripunzip" "rootlesskit" "runc" "rust-bindgen" "rustls-ffi" "s-tui" "sage" "sageWithDoc" "sanoid" "scaphandre" "sftpgo" "sfxr-qt" "sgx-sdk" "silice" "skopeo" "snipe-it" "soft-serve" "sonic-server" "spago" "spark" "spark3" "spark_3_3" "spark_3_4" "spark_3_5" "spdlog" "sqlite" "sqlite-interactive" "sqlite3-to-mysql" "ssm-agent" "ssmsh" "sssd" "staticjinja" "steck" "stig" "stratis-cli" "stratisd" "sudo-rs" "suwayomi-server" "syncthing" "syncthing-cli" "syncthing-discovery" "syncthing-relay" "systemd" "systemd-lock-handler" "systemdLibs" "systemdMinimal" "systemdUkify" "tang" "tectonic" "telegraf" "teleport" "teleport_12" "teleport_13" "teleport_14" "tengine" "termite" "termite-unwrapped" "thanos" "thelounge" "tinyscheme" "tomcat10" "tomcat9" "tracee" "transmission" "transmission-gtk" "transmission-qt" "transmission_4" "transmission_4-gtk" "transmission_4-qt" "transmission_noSystemd" "tsm-client" "tsm-client-withGui" "tuptime" "udev" "unbound" "unbound-full" "unbound-with-systemd" "usbguard" "utf8proc" "varnish" "varnish60" "varnish74" "vault" "vaultwarden" "vaultwarden-mysql" "vaultwarden-postgresql" "vaultwarden-sqlite" "vaultwarden-vault" "vengi-tools" "vips" "vscodium" "vte" "vte-gtk4" "warzone2100" "wasilibc" "wireguard-tools" "wpa_supplicant" "wpa_supplicant_ro_ssids" "wrapGAppsHook" "wrapGAppsHook4" "wrapGAppsNoGuiHook" "xdg-desktop-portal" "xrootd" "xterm" "yarn" "zed" "zlib" "zoneminder" "zq" "zsh" "zstd" ]
```

**Comment by imincik** [comment]
@dotlambda , are you ok with merging this PR as it is now ?

File: pkgs/development/libraries/gdal/tests.nix [hunk 4]
```diff
@@ -1,34 +1,40 @@
...
-    ${gdal}/bin/gdalinfo --version \
-      | grep 'GDAL ${version}'
-
-
-    # test formats
-    ${gdal}/bin/ogrinfo --formats \
+{
+  ogrinfo-version = testers.testVersion {
+    package = gdal;
+    command = "${lib.getExe' gdal "ogrinfo"} --version";
```
Comment (dotlambda) [inline pkgs/development/libraries/gdal/tests.nix:10]:
```suggestion
    command = "ogrinfo --version";
```

File: pkgs/development/libraries/gdal/tests.nix [hunk 5]
```diff
@@ -1,34 +1,40 @@
...
+  };
+
+  gdalinfo-version = testers.testVersion {
+    package = gdal;
+    command = "${lib.getExe' gdal "gdalinfo"} --version";
```
Comment (dotlambda) [inline pkgs/development/libraries/gdal/tests.nix:15]:
```suggestion
    command = "gdalinfo --version";
```

**Pull request merged**